pkg/common-controller/groupsnapshot_controller.go

```diff
@@ -9,3 +22,236 @@ func (ctrl *csiSnapshotCommonController) storeGroupSnapshotUpdate(groupsnapshot
+	}
+
+	return defaultClasses[0], newGroupSnapshot, nil
+}
+
+// pvDriverFromGroupSnapshot is a helper function to get the CSI driver name from the targeted PersistentVolume.
+// It looks up the PVC from which the snapshot is specified to be created from, and looks for the PVC's corresponding
+// PV. Bi-directional binding will be verified between PVC and PV before the PV's CSI driver is returned.
+// For an non-CSI volume, it returns an error immediately as it's not supported.
+func (ctrl *csiSnapshotCommonController) pvDriverFromGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (string, error) {
+	pv, err := ctrl.getVolumeFromVolumeGroupSnapshot(groupSnapshot)
+	if err != nil {
+		return "", err
+	}
+	// supports ONLY CSI volumes
+	if pv.Spec.PersistentVolumeSource.CSI == nil {
+		return "", fmt.Errorf("snapshotting non-CSI volumes is not supported, snapshot:%s/%s", groupSnapshot.Namespace, groupSnapshot.Name)
+	}
+	return pv.Spec.PersistentVolumeSource.CSI.Driver, nil
+}
+
+func (ctrl *csiSnapshotCommonController) getVolumeFromVolumeGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (*v1.PersistentVolume, error) {
+	pvc, err := ctrl.getClaimFromVolumeGroupSnapshot(groupSnapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	if pvc.Status.Phase != v1.ClaimBound {
+		return nil, fmt.Errorf("the PVC %s is not yet bound to a PV, will not attempt to take a group snapshot", pvc.Name)
+	}
+
+	pvName := pvc.Spec.VolumeName
+	pv, err := ctrl.client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve PV %s from the API server: %q", pvName, err)
+	}
+
+	// Verify binding between PV/PVC is still valid
+	bound := ctrl.isVolumeBoundToClaim(pv, pvc)
+	if bound == false {
+		klog.Warningf("binding between PV %s and PVC %s is broken", pvName, pvc.Name)
+		return nil, fmt.Errorf("claim in dataSource not bound or invalid")
+	}
+
+	klog.V(5).Infof("getVolumeFromVolumeGroupSnapshot: group snapshot [%s] PV name [%s]", groupSnapshot.Name, pvName)
+
+	return pv, nil
+}
+
+// getClaimFromVolumeGroupSnapshot is a helper function to get a PVC from VolumeGroupSnapshot.
+func (ctrl *csiSnapshotCommonController) getClaimFromVolumeGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (*v1.PersistentVolumeClaim, error) {
+	labelSelector := groupSnapshot.Spec.Source.Selector
+
+	// Get PVC that has group snapshot label applied.
+	pvcList, err := ctrl.client.CoreV1().PersistentVolumeClaims(groupSnapshot.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector.String()})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list PVCs with label selector %s: %q", labelSelector.String(), err)
+	}
+	if len(pvcList.Items) == 0 {
+		return nil, fmt.Errorf("label selector %s for group snapshot not applied to any PVC", labelSelector.String())
+	}
+	return &pvcList.Items[0], nil
```

Comment:
This only returns the first PVC so it is not useful.  I think we should be able to use existing functions in snapshot controller.

Reply:
Modified to return a list of PVCs. See comment above.